File: backups/PRODUCTION_STABLE_20251112_183253/frontend/src/lib/utils/hebrewCalendar.ts
/**
 * Hebrew Calendar & Biblical Feast Day Calculator
 * Based on HEBREW_FEAST_SYSTEM documentation
 * 
 * Implements accurate Hebrew calendar calculations for 10 biblical feasts:
 * - Spring Feasts: Passover, Unleavened Bread, First Fruits, Pentecost
 * - Fall Feasts: Trumpets, Day of Atonement, Tabernacles, Eighth Day
 * - Other: Purim, Hanukkah
 */

import type { HebrewFeast, FeastProximity } from '../types/celestial'

/**
 * Calculate Passover date using Gauss algorithm
 * Passover falls on Nissan 14 (approximately March/April)
 */
export function calculatePassover(year: number): Date {
  // Gauss Passover algorithm
  const a = (12 * year + 12) % 19
  const b = year % 4
  const Q = -1.904412361576 + 1.554241796621 * a + 0.25 * b - 0.003177794022 * year
  const INTQ = Math.floor(Q)
  const j = (INTQ + 3 * year + 5 * b + 2) % 7
  const r = Q - INTQ
  
  let D: number
  if (j === 2 || j === 4 || j === 6) {
    D = INTQ + 23
  } else if (j === 1 && a > 6 && r >= 0.632870370) {
    D = INTQ + 24
  } else if (j === 0 && a > 11 && r >= 0.897723765) {
    D = INTQ + 23
  } else {
    D = INTQ + 22
  }
  
  // Passover is in March or April
  if (D <= 31) {
    return new Date(year, 2, D) // March
  } else {
    return new Date(year, 3, D - 31) // April
  }
}

/**
 * Calculate Hebrew year from Gregorian date
 */
export function getHebrewYear(date: Date): number {
  const gregorianYear = date.getFullYear()
  const month = date.getMonth()
  
  // Hebrew year starts in September/October
  // If before September, use gregorianYear + 3760
  // If after September, use gregorianYear + 3761
  if (month < 8) {
    return gregorianYear + 3760
  } else {
    return gregorianYear + 3761
  }
}

/**
 * Add days to a date
 */
function addDays(date: Date, days: number): Date {
  const result = new Date(date)
  result.setDate(result.getDate() + days)
  return result
}

/**
 * Get next Sabbath after date (Saturday)
 */
function getNextSabbath(date: Date): Date {
  const result = new Date(date)
  const daysUntilSaturday = (6 - result.getDay() + 7) % 7
  result.setDate(result.getDate() + daysUntilSaturday)
  return result
}

/**
 * Calculate all 10 Hebrew feasts for a given year
 */
export function getHebrewFeastsForYear(year: number): HebrewFeast[] {
  const passover = calculatePassover(year)
  const feasts: HebrewFeast[] = []
  
  // 1. PASSOVER (Pesach) - Nissan 14
  feasts.push({
    name: 'Passover',
    hebrewName: 'Pesach',
    date: passover,
    significance: 'high',
    biblicalReference: 'Leviticus 23:5',
    propheticImportance: 'Commemorates the Exodus; prophetically points to the sacrifice of the Messiah',
    hebrewMonth: 'Nissan',
    hebrewDay: 14,
  })
  
  // 2. UNLEAVENED BREAD - Nissan 15-21 (7 days)
  const unleavenedStart = addDays(passover, 1)
  const unleavenedEnd = addDays(passover, 7)
  feasts.push({
    name: 'Unleavened Bread',
    hebrewName: 'Chag HaMatzot',
    date: unleavenedStart,
    endDate: unleavenedEnd,
    significance: 'high',
    biblicalReference: 'Leviticus 23:6-8',
    propheticImportance: 'Represents putting away sin and living in holiness',
    hebrewMonth: 'Nissan',
    hebrewDay: 15,
  })
  
  // 3. FIRST FRUITS - Day after Sabbath during Unleavened Bread
  const firstSabbathDuringUnleavened = getNextSabbath(unleavenedStart)
  const firstFruits = addDays(firstSabbathDuringUnleavened, 1)
  feasts.push({
    name: 'First Fruits',
    hebrewName: 'Yom HaBikkurim',
    date: firstFruits,
    significance: 'medium',
    biblicalReference: 'Leviticus 23:10-14',
    propheticImportance: 'Prophetically represents the resurrection of the Messiah',
    hebrewMonth: 'Nissan',
    hebrewDay: 16,
  })
  
  // 4. PENTECOST (Shavuot) - 50 days after First Fruits
  const pentecost = addDays(firstFruits, 49)
  feasts.push({
    name: 'Pentecost',
    hebrewName: 'Shavuot',
    date: pentecost,
    significance: 'medium',
    biblicalReference: 'Leviticus 23:15-22',
    propheticImportance: 'Commemorates giving of Torah; prophetically the outpouring of the Holy Spirit',
    hebrewMonth: 'Sivan',
    hebrewDay: 6,
  })
  
  // Calculate fall feasts (Tishrei 1 is approximately 163 days after Passover)
  const trumpets = addDays(passover, 163)
  
  // 5. TRUMPETS (Rosh Hashanah) - Tishrei 1
  feasts.push({
    name: 'Trumpets',
    hebrewName: 'Yom Teruah',
    date: trumpets,
    significance: 'high',
    biblicalReference: 'Leviticus 23:23-25',
    propheticImportance: 'Prophetically represents the rapture and the awakening call',
    hebrewMonth: 'Tishrei',
    hebrewDay: 1,
  })
  
  // 6. DAY OF ATONEMENT (Yom Kippur) - Tishrei 10
  const yomKippur = addDays(trumpets, 9)
  feasts.push({
    name: 'Day of Atonement',
    hebrewName: 'Yom Kippur',
    date: yomKippur,
    significance: 'high',
    biblicalReference: 'Leviticus 23:26-32',
    propheticImportance: 'Holiest day; prophetically represents national Israel\'s repentance and cleansing',
    hebrewMonth: 'Tishrei',
    hebrewDay: 10,
  })
  
  // 7. TABERNACLES (Sukkot) - Tishrei 15-21 (7 days)
  const sukkotStart = addDays(trumpets, 14)
  const sukkotEnd = addDays(trumpets, 20)
  feasts.push({
    name: 'Tabernacles',
    hebrewName: 'Sukkot',
    date: sukkotStart,
    endDate: sukkotEnd,
    significance: 'high',
    biblicalReference: 'Leviticus 23:33-43',
    propheticImportance: 'Prophetically represents the Millennial Kingdom reign',
    hebrewMonth: 'Tishrei',
    hebrewDay: 15,
  })
  
  // 8. EIGHTH DAY (Shemini Atzeret) - Tishrei 22
  const eighthDay = addDays(trumpets, 21)
  feasts.push({
    name: 'Eighth Day',
    hebrewName: 'Shemini Atzeret',
    date: eighthDay,
    significance: 'medium',
    biblicalReference: 'Leviticus 23:36',
    propheticImportance: 'Represents the eternal state and New Jerusalem',
    hebrewMonth: 'Tishrei',
    hebrewDay: 22,
  })
  
  // 9. PURIM - Adar 14 (approximately February/March)
  // Purim is 30 days before Passover
  const purim = addDays(passover, -30)
  feasts.push({
    name: 'Purim',
    hebrewName: 'Purim',
    date: purim,
    significance: 'low',
    biblicalReference: 'Esther 9:26-28',
    propheticImportance: 'Celebrates deliverance from Haman; represents God\'s providence',
    hebrewMonth: 'Adar',
    hebrewDay: 14,
  })
  
  // 10. HANUKKAH (Dedication) - Kislev 25 - Tevet 2 (8 days)
  // Hanukkah is approximately 75 days before Passover
  const hanukkahStart = addDays(passover, -75)
  const hanukkahEnd = addDays(hanukkahStart, 7)
  feasts.push({
    name: 'Hanukkah',
    hebrewName: 'Chanukah',
    date: hanukkahStart,
    endDate: hanukkahEnd,
    significance: 'low',
    biblicalReference: 'John 10:22-23',
    propheticImportance: 'Festival of Lights; represents rededication and spiritual victory',
    hebrewMonth: 'Kislev',
    hebrewDay: 25,
  })
  
  return feasts
}

/**
 * Check if a date is within proximity of any Hebrew feast
 * Default tolerance: ±3 days
 */
export function getFeastProximity(date: Date, toleranceDays: number = 3): FeastProximity | null {
  const year = date.getFullYear()
  const feasts = getHebrewFeastsForYear(year)
  
  // Also check previous and next year for boundary cases
  const prevYearFeasts = getHebrewFeastsForYear(year - 1)
  const nextYearFeasts = getHebrewFeastsForYear(year + 1)
  const allFeasts = [...prevYearFeasts, ...feasts, ...nextYearFeasts]
  
  for (const feast of allFeasts) {
    // Check main date
    const daysDiff = Math.abs(differenceInDays(date, feast.date))
    if (daysDiff <= toleranceDays) {
      return {
        feast,
        daysAway: differenceInDays(feast.date, date),
        isExactMatch: daysDiff === 0,
        isWithinTolerance: true,
      }
    }
    
    // Check end date for multi-day feasts
    if (feast.endDate) {
      const isWithinRange = date >= feast.date && date <= feast.endDate
      if (isWithinRange) {
        return {
          feast,
          daysAway: 0,
          isExactMatch: true,
          isWithinTolerance: true,
        }
      }
    }
  }
  
  return null
}

/**
 * Get all feasts within a date range
 */
export function getFeastsInRange(startDate: Date, endDate: Date): HebrewFeast[] {
  const startYear = startDate.getFullYear()
  const endYear = endDate.getFullYear()
  const allFeasts: HebrewFeast[] = []
  
  for (let year = startYear; year <= endYear; year++) {
    const yearFeasts = getHebrewFeastsForYear(year)
    const feastsInRange = yearFeasts.filter(feast => {
      const feastDate = feast.date
      return feastDate >= startDate && feastDate <= endDate
    })
    allFeasts.push(...feastsInRange)
  }
  
  return allFeasts
}

/**
 * Format feast date range for display
 */
export function formatFeastDateRange(feast: HebrewFeast): string {
  const options: Intl.DateTimeFormatOptions = { month: 'short', day: 'numeric', year: 'numeric' }
  const startStr = feast.date.toLocaleDateString('en-US', options)
  
  if (feast.endDate) {
    const endStr = feast.endDate.toLocaleDateString('en-US', options)
    return `${startStr} - ${endStr}`
  }
  
  return startStr
}

/**
 * Calculate difference in days between two dates
 */
function differenceInDays(date1: Date, date2: Date): number {
  const msPerDay = 1000 * 60 * 60 * 24
  const utc1 = Date.UTC(date1.getFullYear(), date1.getMonth(), date1.getDate())
  const utc2 = Date.UTC(date2.getFullYear(), date2.getMonth(), date2.getDate())
  return Math.floor((utc1 - utc2) / msPerDay)
}

File: backend/app/integrations/hebrew_calendar.py
"""
Hebrew Calendar Integration
Calculates Jewish feast days using the Hebrew calendar

Biblical Feasts (Moedim):
- Passover (Pesach) - Nisan 14
- Feast of Unleavened Bread - Nisan 15-21
- Pentecost (Shavuot) - 50 days after Passover
- Feast of Trumpets (Rosh Hashanah) - Tishrei 1
- Day of Atonement (Yom Kippur) - Tishrei 10
- Feast of Tabernacles (Sukkot) - Tishrei 15-21
"""

from datetime import datetime, timedelta
from typing import List, Dict, Tuple
import convertdate
from convertdate import hebrew


class HebrewCalendar:
    """Hebrew calendar calculations for biblical feast days"""
    
    # Hebrew month names
    NISAN = 1   # Spring month (March/April)
    IYAR = 2
    SIVAN = 3
    TAMMUZ = 4
    AV = 5
    ELUL = 6
    TISHREI = 7  # Fall month (September/October)
    CHESHVAN = 8
    KISLEV = 9
    TEVET = 10
    SHEVAT = 11
    ADAR = 12
    
    @staticmethod
    def gregorian_to_hebrew(year: int, month: int, day: int) -> Tuple[int, int, int]:
        """Convert Gregorian date to Hebrew calendar"""
        return hebrew.from_gregorian(year, month, day)
    
    @staticmethod
    def hebrew_to_gregorian(year: int, month: int, day: int) -> Tuple[int, int, int]:
        """Convert Hebrew date to Gregorian calendar"""
        return hebrew.to_gregorian(year, month, day)
    
    @classmethod
    def calculate_passover(cls, gregorian_year: int) -> datetime:
        """
        Calculate Passover (Pesach) - Nisan 14
        
        Args:
            gregorian_year: Gregorian year (e.g., 2025)
            
        Returns:
            datetime: Passover date in Gregorian calendar
        """
        # Get Hebrew year corresponding to spring of Gregorian year
        # Nisan is in spring, so Hebrew year starts in fall of previous Gregorian year
        hebrew_year = gregorian_year + 3760
        
        # Passover is Nisan 14
        greg_year, greg_month, greg_day = cls.hebrew_to_gregorian(hebrew_year, cls.NISAN, 14)
        
        # If Passover falls before March, it's actually in next Hebrew year
        if greg_month < 3:
            hebrew_year += 1
            greg_year, greg_month, greg_day = cls.hebrew_to_gregorian(hebrew_year, cls.NISAN, 14)
            
        return datetime(greg_year, greg_month, greg_day)
    
    @classmethod
    def calculate_unleavened_bread(cls, gregorian_year: int) -> Tuple[datetime, datetime]:
        """
        Calculate Feast of Unleavened Bread - Nisan 15-21
        
        Returns:
            Tuple[datetime, datetime]: (start_date, end_date)
        """
        passover = cls.calculate_passover(gregorian_year)
        start_date = passover + timedelta(days=1)  # Nisan 15
        end_date = start_date + timedelta(days=6)  # Nisan 21
        return (start_date, end_date)
    
    @classmethod
    def calculate_pentecost(cls, gregorian_year: int) -> datetime:
        """
        Calculate Pentecost (Shavuot) - 50 days after Passover
        Counted from the day after the Sabbath during Unleavened Bread
        
        Args:
            gregorian_year: Gregorian year
            
        Returns:
            datetime: Pentecost date
        """
        passover = cls.calculate_passover(gregorian_year)
        # Count 50 days from the day after the Sabbath
        # Simplified: 50 days from Nisan 16 (day after Passover)
        pentecost = passover + timedelta(days=50)
        return pentecost
    
    @classmethod
    def calculate_trumpets(cls, gregorian_year: int) -> datetime:
        """
        Calculate Feast of Trumpets (Rosh Hashanah) - Tishrei 1
        This marks the beginning of the civil Hebrew year
        
        Args:
            gregorian_year: Gregorian year
            
        Returns:
            datetime: Feast of Trumpets date
        """
        # Tishrei 1 falls in fall of Gregorian year
        # Hebrew year for fall feasts is gregorian_year + 3761
        hebrew_year = gregorian_year + 3761
        
        greg_year, greg_month, greg_day = cls.hebrew_to_gregorian(hebrew_year, cls.TISHREI, 1)
        
        # If Tishrei falls after December, use previous Hebrew year
        if greg_month > 12:
            hebrew_year -= 1
            greg_year, greg_month, greg_day = cls.hebrew_to_gregorian(hebrew_year, cls.TISHREI, 1)
            
        return datetime(greg_year, greg_month, greg_day)
    
    @classmethod
    def calculate_atonement(cls, gregorian_year: int) -> datetime:
        """
        Calculate Day of Atonement (Yom Kippur) - Tishrei 10
        
        Args:
            gregorian_year: Gregorian year
            
        Returns:
            datetime: Day of Atonement date
        """
        trumpets = cls.calculate_trumpets(gregorian_year)
        # Yom Kippur is 9 days after Rosh Hashanah (Tishrei 10)
        atonement = trumpets + timedelta(days=9)
        return atonement
    
    @classmethod
    def calculate_tabernacles(cls, gregorian_year: int) -> Tuple[datetime, datetime]:
        """
        Calculate Feast of Tabernacles (Sukkot) - Tishrei 15-21
        
        Returns:
            Tuple[datetime, datetime]: (start_date, end_date)
        """
        trumpets = cls.calculate_trumpets(gregorian_year)
        start_date = trumpets + timedelta(days=14)  # Tishrei 15
        end_date = start_date + timedelta(days=6)   # Tishrei 21
        return (start_date, end_date)
    
    @classmethod
    def get_all_feasts(cls, gregorian_year: int) -> Dict[str, Dict]:
        """
        Get all biblical feast days for a given Gregorian year
        
        Args:
            gregorian_year: Gregorian year (e.g., 2025)
            
        Returns:
            Dict: Dictionary of feast names to date information
        """
        unleavened = cls.calculate_unleavened_bread(gregorian_year)
        tabernacles = cls.calculate_tabernacles(gregorian_year)
        
        return {
            "passover": {
                "name": "Passover (Pesach)",
                "date": cls.calculate_passover(gregorian_year),
                "hebrew_date": "Nisan 14",
                "is_range": False,
                "significance": "Commemorates the Exodus from Egypt"
            },
            "unleavened_bread": {
                "name": "Feast of Unleavened Bread",
                "start_date": unleavened[0],
                "end_date": unleavened[1],
                "hebrew_date": "Nisan 15-21",
                "is_range": True,
                "significance": "Commemorates the haste of the Exodus"
            },
            "pentecost": {
                "name": "Pentecost (Shavuot)",
                "date": cls.calculate_pentecost(gregorian_year),
                "hebrew_date": "Sivan 6 (50 days after Passover)",
                "is_range": False,
                "significance": "Commemorates the giving of the Torah at Mount Sinai"
            },
            "trumpets": {
                "name": "Feast of Trumpets (Rosh Hashanah)",
                "date": cls.calculate_trumpets(gregorian_year),
                "hebrew_date": "Tishrei 1",
                "is_range": False,
                "significance": "Jewish New Year, Day of Judgment"
            },
            "atonement": {
                "name": "Day of Atonement (Yom Kippur)",
                "date": cls.calculate_atonement(gregorian_year),
                "hebrew_date": "Tishrei 10",
                "is_range": False,
                "significance": "Holiest day, day of repentance and forgiveness"
            },
            "tabernacles": {
                "name": "Feast of Tabernacles (Sukkot)",
                "start_date": tabernacles[0],
                "end_date": tabernacles[1],
                "hebrew_date": "Tishrei 15-21",
                "is_range": True,
                "significance": "Commemorates the 40 years in the wilderness"
            }
        }
    
    @classmethod
    def get_feast_years_range(cls, start_year: int, end_year: int) -> List[Dict]:
        """
        Get all biblical feast days for a range of years
        
        Args:
            start_year: Starting Gregorian year
            end_year: Ending Gregorian year
            
        Returns:
            List[Dict]: List of feast day records across all years
        """
        all_feasts = []
        
        for year in range(start_year, end_year + 1):
            year_feasts = cls.get_all_feasts(year)
            
            for feast_key, feast_data in year_feasts.items():
                feast_record = {
                    "year": year,
                    "feast_type": feast_key,
                    "name": feast_data["name"],
                    "hebrew_date": feast_data["hebrew_date"],
                    "significance": feast_data["significance"],
                    "is_range": feast_data["is_range"]
                }
                
                if feast_data["is_range"]:
                    feast_record["start_date"] = feast_data["start_date"]
                    feast_record["end_date"] = feast_data["end_date"]
                else:
                    feast_record["date"] = feast_data["date"]
                    
                all_feasts.append(feast_record)
        
        return all_feasts


# Convenience functions for direct API use
def get_passover(year: int) -> datetime:
    """Get Passover date for a given year"""
    return HebrewCalendar.calculate_passover(year)


def get_pentecost(year: int) -> datetime:
    """Get Pentecost date for a given year"""
    return HebrewCalendar.calculate_pentecost(year)


def get_trumpets(year: int) -> datetime:
    """Get Feast of Trumpets date for a given year"""
    return HebrewCalendar.calculate_trumpets(year)


def get_atonement(year: int) -> datetime:
    """Get Day of Atonement date for a given year"""
    return HebrewCalendar.calculate_atonement(year)


def get_tabernacles(year: int) -> Tuple[datetime, datetime]:
    """Get Feast of Tabernacles dates for a given year"""
    return HebrewCalendar.calculate_tabernacles(year)


def get_all_feasts_for_year(year: int) -> Dict[str, Dict]:
    """Get all feast days for a single year"""
    return HebrewCalendar.get_all_feasts(year)


def get_feasts_range(start_year: int, end_year: int) -> List[Dict]:
    """Get all feast days across a range of years"""
    return HebrewCalendar.get_feast_years_range(start_year, end_year)


if __name__ == "__main__":
    # Test the calculations
    print("📅 Hebrew Calendar - Biblical Feast Days Calculator\n")
    
    test_year = 2025
    print(f"Biblical Feasts for {test_year}:\n")
    
    feasts = HebrewCalendar.get_all_feasts(test_year)
    
    for feast_key, feast_data in feasts.items():
        print(f"🕎 {feast_data['name']}")
        print(f"   Hebrew Date: {feast_data['hebrew_date']}")
        
        if feast_data['is_range']:
            print(f"   Gregorian: {feast_data['start_date'].strftime('%B %d, %Y')} - {feast_data['end_date'].strftime('%B %d, %Y')}")
        else:
            print(f"   Gregorian: {feast_data['date'].strftime('%B %d, %Y')}")
            
        print(f"   Significance: {feast_data['significance']}\n")
